sympy/matrices/sparse.py

```diff
@@ -61,32 +135,78 @@ def __new__(cls, *args, **kwargs):
                         if value:
                             self._smat[(i, j)] = value
             elif isinstance(args[2], (dict, Dict)):
-                # manual copy, copy.deepcopy() doesn't work
-                for key in args[2].keys():
-                    v = args[2][key]
+                def update(i, j, v):
+                    # update self._smat and make sure there are
+                    # no collisions
                     if v:
-                        self._smat[key] = self._sympify(v)
+                        if (i, j) in self._smat and v != self._smat[i, j]:
+                            raise ValueError('collision at %s' % ((i, j),))
+                        self._smat[i, j] = v
+                # manual copy, copy.deepcopy() doesn't work
+                for key, v in args[2].items():
+                    r, c = key
+                    if isinstance(v, SparseMatrix):
+                        for (i, j), vij in v._smat.items():
+                            update(r + i, c + j, vij)
+                    else:
+                        if isinstance(v, (Matrix, list, tuple)):
+                            v = SparseMatrix(v)
+                            for i, j in v._smat:
+                                update(r + i, c + j, v[i, j])
+                        else:
+                            v = self._sympify(v)
+                            update(r, c, self._sympify(v))
             elif is_sequence(args[2]):
-                if len(args[2]) != self.rows*self.cols:
-                    raise ValueError(
-                        'List length (%s) != rows*columns (%s)' %
-                        (len(args[2]), self.rows*self.cols))
-                flat_list = args[2]
+                flat = not any(is_sequence(i) for i in args[2])
```

Comment:
Maybe call it `notflat` to avoid double negation? (Or change order of the if-else clauses below?)

Reply:
I will change this in another PR to avoid a test run. Perhaps I can call it 'nested = any(is_sequence(i) for i in args[2])` and test `if nested` below.